autogo/shadow.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// autogo creates a shadow copy of the go packages at origin in a destination.
+//
+// In other words this program will walk the directory tree at origin
+// and for each:
+// * directory - create a directory with the same name in destination
+// * go-related-file - link the file into the matching destination directory
+//
+// The effect is similar to:
+//   rsync -zarv --include="*/" --include="*.sh" --exclude="*" "$from" "$to"
+// TODO(fejta): investigate just using rsync?
+//
+// The intended use case of this program is with the autogo_generate in //autogo:def.bzl. This rule will clone your primary workspace into an autogo workspace, and then run gazelle to generate rules for go packages.
+//
+// Usage:
+//   autogo -- <ORIGIN_DIR> <DEST_DIR>
+//
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// main calls shadowClone and exits non-zero on failure
+func main() {
+	if n := len(os.Args); n != 3 {
+		log.Printf("Expected 2 args, not %d", n)
+		log.Fatalf("Usage: %s <ORIGIN> <DEST>", filepath.Base(os.Args[0]))
+	}
+	if err := shadowClone(os.Args[1], os.Args[2]); err != nil {
+		log.Fatalf("Failed to clone %s to %s: %v", os.Args[1], os.Args[2], err)
+	}
+}
+
+// shadowClone walks origin to clone the directory structure and link files at the desination.
+func shadowClone(origin, dest string) error {
+	v := visitor{
+		origin: origin,
+		dest:   dest,
+	}
+	return filepath.Walk(v.origin, v.visit)
+}
+
+// action does something with the current file (mkdir or link)
+type action func(origin string, info os.FileInfo, dest string) error
+
+// visitor stores the origin we're cloning from and destionation we clone to.
+type visitor struct {
+	// origin is the path we read to determine what to clone
+	origin string
+	// dest is where we clone
+	dest string
+}
+
+// visit chooses and then performs the right action for the current path
+func (v visitor) visit(path string, info os.FileInfo, verr error) error {
+	act, dest, err := v.choose(path, info, verr)
+	if err != nil {
+		return err
+	}
+	if act == nil {
+		return nil
+	}
+	return act(path, info, dest)
+}
+
+// choose looks at the current path and returns the appropriate action:
+// - testdata directory => skip it and everything under it
+// - directory => create the directory
+// - go file (.go, .s) => create it at dest
+// - other files => do nothing
```

Comment:
gazelle handles proto files - we probably should include those too/

Reply:
Double check that proto files are buildable with Bazel first. There are some restrictions on import paths.

Note that gazelle will ignore proto files in vendor directories by default.

Reply:
Added a TODO